weld/vectorizer.rs

```diff
@@ -14,12 +14,42 @@ use std::collections::HashSet;
 
 use super::exprs;
 
-/// Vectorizes a type.
+/// Get the vectorized version of a type.
 fn vectorized_type(ty: &Type) -> Type {
-    if let Scalar(kind) = *ty {
-        Simd(kind)
-    } else {
-        ty.clone()
+    match *ty {
+        Scalar(kind) => Simd(kind),
+        Struct(ref fields) => Struct(fields.iter().map(|ref f| vectorized_type(f)).collect()),
+        _ => ty.clone()
+    }
+}
+
+/// Return true if a given type is SIMD vectorized, which concretely means that it is
+/// either a simd[T] or a struct of SIMD types.
+///
+/// TODO: we should investigate just using simd[struct[T]] for structs to turn this into
+/// a much simpler pattern match.
+pub fn is_simd(ty: &Type) -> bool {
+    match *ty {
+        Simd(_) => true,
+        Struct(ref fields) => fields.iter().all(is_simd),
+        _ => false
+    }
+}
+
+/// Get the scalar type corresponding to a SIMD type.
+pub fn scalar_type(simd_type: &Type) -> WeldResult<Type> {
```

Comment:
Because `vectorized_type` was here. It may be OK to move them, but then we should move all 3 and we would probably want to make `vectorized_type` to return a WeldResult too to be consistent.